### PR TITLE
workflow continuation after app restart

### PR DESCRIPTION
**Pull request by RichardW98**
task entity

Screenshot 2022-12-21 at 12 41 59 PM

### CODE REVIEW

File: services/infrastructure-service/pom.xml [hunk 1]
```diff
@@ -1,174 +1,181 @@
...
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <!-- START Test Dependencies -->
```
Comment (anludke) [inline services/infrastructure-service/pom.xml:144]: END

File: services/infrastructure-service/pom.xml [hunk 2]
```diff
@@ -1,174 +1,181 @@
...
+
+        <!-- START Paging Results Dependencies -->
+        <dependency>
+            <groupId>org.springframework.hateoas</groupId>
+            <artifactId>spring-hateoas</artifactId>
+        </dependency>
+        <!-- END Paging Results Dependencies -->
+
+        <!-- START AOP -->
```
Comment (anludke) [inline services/infrastructure-service/pom.xml:153]: START AOP Dependencies

File: services/infrastructure-service/pom.xml [hunk 3]
```diff
@@ -1,174 +1,181 @@
...
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-aop</artifactId>
+        </dependency>
+        <!-- START AOP -->
```
Comment (anludke) [inline services/infrastructure-service/pom.xml:158]: END AOP Dependencies

**Review by anludke — CHANGES_REQUESTED** [review]

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/TaskAspect.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            if (isExceptionThrown)
+                workFlowTransactionEntity.setStatus(WorkFlowStatus.FAILED.name());
+        }
+        return workTransactionService.updateWorkFlowTransactionEntity(workFlowTransactionEntity);
+    }
+
+    private WorkFlowStatus getTaskStatus(WorkFlowTransactionEntity workFlowTransactionEntity, String taskName) {
+        log.info("checking task status: {} in Workflow: {}", taskName, workFlowTransactionEntity.getId());
+        return workFlowTransactionEntity.getTaskTransactions().stream()
+                .filter(taskTransactionEntity -> taskName.equals(taskTransactionEntity.getTaskName()))
```
Comment (anludke) [inline services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/TaskAspect.java:130]: It should be better to combine (and) both predicates use in filter into one.

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @author Richard Wang (Github: RichardW98)
+ */
+@Data
+@EqualsAndHashCode(callSuper = false)
+@Entity(name = "task_transaction")
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class TaskTransactionEntity extends AbstractEntity {
```
Comment (anludke) [inline services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java:43]: TaskTransactionEntity <> WorkFlowTransactionEntity?

**Review by anludke — CHANGES_REQUESTED** [review]

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    @ManyToOne
+    @JoinColumn(name = "workflow_transaction_id", nullable = false)
+    private WorkFlowTransactionEntity workFlowTransaction;
+
+    private String taskName;
+
+    private WorkFlowStatus taskStatus;
```
Comment (anludke) [inline services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java:50]: Let's have TaskStatus that's extends WorkFlowStatus

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+@Data
+@EqualsAndHashCode(callSuper = false)
+@Entity(name = "task_transaction")
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class TaskTransactionEntity extends AbstractEntity {
+    @ManyToOne
+    @JoinColumn(name = "workflow_transaction_id", nullable = false)
+    private WorkFlowTransactionEntity workFlowTransaction;
```
Comment (anludke) [inline services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java:46]: let's rename to workFlowTransactionEntity

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/TaskAspect.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    public WorkReport executeAroundAdvice(ProceedingJoinPoint proceedingJoinPoint, WorkContext workContext) {
+        String taskName = ((WorkFlowTask)proceedingJoinPoint.getTarget()).getName();
+        WorkFlowTransactionEntity workFlowTransactionEntity = workTransactionService.getWorkFlowTransactionEntity(String.valueOf(workContext.get(WorkFlowConstants.WORKFLOW_EXECUTION_ENTITY_REFERENCES)));
+        log.info("Before invoking execute() on task: {}", taskName);
+        WorkFlowStatus status = getTaskStatus(workFlowTransactionEntity, taskName);
+        if (status != null) {
+            log.info("task status found: {}", taskName);
+            return new DefaultWorkReport(status.equals(WorkFlowStatus.SUCCESS) ? WorkStatus.COMPLETED : WorkStatus.FAILED, workContext);
+        }
+        workFlowTransactionEntity = updateTaskPersistence(workFlowTransactionEntity, taskName, workContext, WorkFlowStatus.IN_PROGRESS, true, false);
```
Comment (anludke) [inline services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/TaskAspect.java:71]: only add task transaction records when it's the first time

File: services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java (hunk 6 above)
Comment (RichardW98) [reply services/infrastructure-service/src/main/java/com/redhat/parodos/workflow/execution/transaction/TaskTransactionEntity.java]: renamed to Status

**Review by anludke — APPROVED** [review]

**Pull request merged**